File: frontend/src/api/rfi.ts
import { apiClient } from './client'

export interface RFI {
  id: string
  projectId: string
  rfiNumber: string
  emailThreadId?: string
  emailMessageId?: string
  subject: string
  question: string
  category: string
  priority: string
  toEmail: string
  toName?: string
  ccEmails?: string[]
  status: string
  dueDate?: string
  respondedAt?: string
  closedAt?: string
  location?: string
  drawingReference?: string
  specificationReference?: string
  attachments?: Record<string, unknown>[]
  createdAt: string
  updatedAt: string
  sentAt?: string
  createdBy?: {
    id: string
    email: string
    fullName: string
  }
  assignedTo?: {
    id: string
    email: string
    fullName: string
  }
  relatedEquipmentId?: string
  relatedMaterialId?: string
  relatedAreaId?: string
  relatedEquipment?: { id: string; name: string; equipmentType?: string }
  relatedMaterial?: { id: string; name: string; materialType?: string }
  relatedArea?: { id: string; name: string; areaCode?: string; floorNumber?: number }
  responses?: RFIResponseData[]
  syncSource?: string
  syncStatus?: string
  lastSyncedAt?: string
}

export interface RFIResponseData {
  id: string
  rfiId: string
  emailMessageId?: string
  inReplyTo?: string
  responseText: string
  attachments?: Array<{ id: string; filename: string; url: string }>
  fromEmail: string
  fromName?: string
  responder?: {
    id: string
    email: string
    fullName: string
  }
  isInternal: boolean
  isCcParticipant: boolean
  source: string
  createdAt: string
  receivedAt?: string
}

export interface RFIListItem {
  id: string
  projectId: string
  rfiNumber: string
  subject: string
  toEmail: string
  toName?: string
  category: string
  priority: string
  status: string
  dueDate?: string
  createdAt: string
  sentAt?: string
  respondedAt?: string
  responseCount: number
  syncSource?: string
  syncStatus?: string
  lastSyncedAt?: string
}

export interface RFICreate {
  subject: string
  question: string
  to_email: string
  to_name?: string
  cc_emails?: string[]
  category?: string
  priority?: string
  due_date?: string
  location?: string
  drawing_reference?: string
  specification_reference?: string
  attachments?: Record<string, unknown>[]
  assigned_to_id?: string
  related_equipment_id?: string
  related_material_id?: string
  related_area_id?: string
}

export interface RFIUpdate {
  subject?: string
  question?: string
  to_email?: string
  to_name?: string
  cc_emails?: string[]
  category?: string
  priority?: string
  due_date?: string
  location?: string
  drawing_reference?: string
  specification_reference?: string
  attachments?: Record<string, unknown>[]
  assigned_to_id?: string
  related_equipment_id?: string
  related_material_id?: string
  related_area_id?: string
}

export interface RFIResponseCreate {
  response_text: string
  attachments?: Record<string, unknown>[]
  is_internal?: boolean
}

export interface RFISummary {
  totalRfis: number
  draftCount: number
  openCount: number
  waitingResponseCount: number
  answeredCount: number
  closedCount: number
  overdueCount: number
  byPriority: Record<string, number>
  byCategory: Record<string, number>
}

export interface PaginatedRFIResponse {
  items: RFIListItem[]
  total: number
  page: number
  pageSize: number
  totalPages: number
}

export interface RFIEmailLog {
  id: string
  rfiId?: string
  eventType: string
  emailMessageId?: string
  fromEmail?: string
  toEmail?: string
  subject?: string
  errorMessage?: string
  createdAt: string
}

export interface RFIFilters {
  status?: string
  priority?: string
  search?: string
  page?: number
  page_size?: number
}

export interface ACCSyncStatus {
  syncHealth: 'ok' | 'warning' | 'error'
  lastSyncedAt?: string
  totalSynced: number
  conflictCount: number
  errorMessage?: string
}

export interface ACCConflictItem {
  id: string
  rfiNumber: string
  subject: string
  conflictFields: string[]
  localUpdatedAt: string
  accUpdatedAt: string
  accMetadata?: Record<string, unknown>
}

export interface ACCConflictResolveRequest {
  strategy: 'last_write_wins' | 'prefer_local' | 'prefer_acc'
}

export const rfiApi = {
  list: async (projectId: string, filters?: RFIFilters): Promise<PaginatedRFIResponse> => {
    const params = new URLSearchParams()
    if (filters?.status) params.append('status', filters.status)
    if (filters?.priority) params.append('priority', filters.priority)
    if (filters?.search) params.append('search', filters.search)
    if (filters?.page) params.append('page', filters.page.toString())
    if (filters?.page_size) params.append('page_size', filters.page_size.toString())

    const url = `/projects/${projectId}/rfis${params.toString() ? `?${params}` : ''}`
    const response = await apiClient.get(url)
    return response.data
  },

  get: async (rfiId: string): Promise<RFI> => {
    const response = await apiClient.get(`/rfis/${rfiId}`)
    return response.data
  },

  create: async (projectId: string, data: RFICreate): Promise<RFI> => {
    const response = await apiClient.post(`/projects/${projectId}/rfis`, data)
    return response.data
  },

  update: async (rfiId: string, data: RFIUpdate): Promise<RFI> => {
    const response = await apiClient.patch(`/rfis/${rfiId}`, data)
    return response.data
  },

  send: async (rfiId: string): Promise<RFI> => {
    const response = await apiClient.post(`/rfis/${rfiId}/send`)
    return response.data
  },

  updateStatus: async (rfiId: string, status: string): Promise<RFI> => {
    const response = await apiClient.patch(`/rfis/${rfiId}/status`, { status })
    return response.data
  },

  addResponse: async (rfiId: string, data: RFIResponseCreate, sendEmail = true): Promise<RFIResponseData> => {
    const response = await apiClient.post(`/rfis/${rfiId}/responses?send_email=${sendEmail}`, data)
    return response.data
  },

  getResponses: async (rfiId: string): Promise<RFIResponseData[]> => {
    const response = await apiClient.get(`/rfis/${rfiId}/responses`)
    return response.data
  },

  getEmailLog: async (rfiId: string): Promise<RFIEmailLog[]> => {
    const response = await apiClient.get(`/rfis/${rfiId}/email-log`)
    return response.data
  },

  getSummary: async (projectId: string): Promise<RFISummary> => {
    const response = await apiClient.get(`/projects/${projectId}/rfis/summary`)
    return response.data
  },

  delete: async (rfiId: string): Promise<void> => {
    await apiClient.delete(`/rfis/${rfiId}`)
  },

  closeRfi: async (rfiId: string): Promise<RFI> => {
    const response = await apiClient.patch(`/rfis/${rfiId}/status`, { status: 'closed' })
    return response.data
  },

  reopenRfi: async (rfiId: string): Promise<RFI> => {
    const response = await apiClient.patch(`/rfis/${rfiId}/status`, { status: 'open' })
    return response.data
  },

  triggerAccSync: async (projectId: string): Promise<{ message: string }> => {
    const response = await apiClient.post(`/projects/${projectId}/acc/sync`)
    return response.data
  },

  getAccSyncStatus: async (projectId: string): Promise<ACCSyncStatus> => {
    const response = await apiClient.get(`/projects/${projectId}/acc/sync/status`)
    return response.data
  },

  getAccConflicts: async (projectId: string): Promise<ACCConflictItem[]> => {
    const response = await apiClient.get(`/projects/${projectId}/acc/conflicts`)
    return response.data
  },

  resolveConflict: async (projectId: string, rfiId: string, strategy: ACCConflictResolveRequest['strategy']): Promise<RFI> => {
    const response = await apiClient.post(`/projects/${projectId}/acc/conflicts/${rfiId}/resolve`, { strategy })
    return response.data
  },
}

export const RFI_STATUS_OPTIONS = [
  { value: 'draft', labelKey: 'rfis.statuses.draft', color: 'default' },
  { value: 'open', labelKey: 'rfis.statuses.open', color: 'info' },
  { value: 'waiting_response', labelKey: 'rfis.statuses.waiting_response', color: 'warning' },
  { value: 'answered', labelKey: 'rfis.statuses.answered', color: 'success' },
  { value: 'closed', labelKey: 'rfis.statuses.closed', color: 'default' },
  { value: 'cancelled', labelKey: 'rfis.statuses.cancelled', color: 'error' },
] as const

export const RFI_PRIORITY_OPTIONS = [
  { value: 'low', labelKey: 'rfis.priorities.low', color: 'default' },
  { value: 'medium', labelKey: 'rfis.priorities.medium', color: 'info' },
  { value: 'high', labelKey: 'rfis.priorities.high', color: 'warning' },
  { value: 'urgent', labelKey: 'rfis.priorities.urgent', color: 'error' },
] as const

export const RFI_CATEGORY_OPTIONS = [
  { value: 'design', labelKey: 'rfis.categories.design' },
  { value: 'structural', labelKey: 'rfis.categories.structural' },
  { value: 'mep', labelKey: 'rfis.categories.mep' },
  { value: 'architectural', labelKey: 'rfis.categories.architectural' },
  { value: 'specifications', labelKey: 'rfis.categories.specifications' },
  { value: 'schedule', labelKey: 'rfis.categories.schedule' },
  { value: 'cost', labelKey: 'rfis.categories.cost' },
  { value: 'other', labelKey: 'rfis.categories.other' },
] as const

File: backend/app/schemas/rfi.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.core.validators import (
    MAX_NAME_LENGTH,
    MAX_NOTES_LENGTH,
    MIN_NAME_LENGTH,
    sanitize_string,
)
from app.schemas.user import UserResponse

MAX_SUBJECT_LENGTH = 500
MAX_EMAIL_LENGTH = 255


class RFIBase(BaseModel):
    subject: str = Field(min_length=MIN_NAME_LENGTH, max_length=MAX_SUBJECT_LENGTH)
    question: str = Field(min_length=MIN_NAME_LENGTH, max_length=MAX_NOTES_LENGTH)
    to_email: EmailStr
    to_name: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    cc_emails: list[EmailStr] | None = Field(default=None, max_length=20)
    category: str = Field(default="other", max_length=50)
    priority: str = Field(default="medium", max_length=20)
    due_date: datetime | None = None
    location: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    drawing_reference: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    specification_reference: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    attachments: list[dict] | None = Field(default=None, max_length=50)

    @field_validator("due_date", mode="before")
    @classmethod
    def strip_timezone(cls, v):
        if v is None:
            return v
        if isinstance(v, str):
            v = datetime.fromisoformat(v)
        if isinstance(v, datetime) and v.tzinfo is not None:
            return v.replace(tzinfo=None)
        return v

    @field_validator("subject", "question", "to_name", "location", mode="before")
    @classmethod
    def sanitize_text(cls, v: str | None) -> str | None:
        return sanitize_string(v)

    @field_validator("category", mode="before")
    @classmethod
    def validate_category(cls, v: str) -> str:
        valid_categories = [
            "design", "structural", "mep", "architectural",
            "specifications", "schedule", "cost", "other"
        ]
        if v and v.lower() not in valid_categories:
            return "other"
        return v.lower() if v else "other"

    @field_validator("priority", mode="before")
    @classmethod
    def validate_priority(cls, v: str) -> str:
        valid_priorities = ["low", "medium", "high", "urgent"]
        if v and v.lower() not in valid_priorities:
            return "medium"
        return v.lower() if v else "medium"


class RFICreate(RFIBase):
    assigned_to_id: UUID | None = None
    related_equipment_id: UUID | None = None
    related_material_id: UUID | None = None


class RFIUpdate(BaseModel):
    subject: str | None = Field(default=None, min_length=MIN_NAME_LENGTH, max_length=MAX_SUBJECT_LENGTH)
    question: str | None = Field(default=None, min_length=MIN_NAME_LENGTH, max_length=MAX_NOTES_LENGTH)
    to_email: EmailStr | None = None
    to_name: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    cc_emails: list[EmailStr] | None = Field(default=None, max_length=20)
    category: str | None = Field(default=None, max_length=50)
    priority: str | None = Field(default=None, max_length=20)
    due_date: datetime | None = None
    location: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    drawing_reference: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    specification_reference: str | None = Field(default=None, max_length=MAX_NAME_LENGTH)
    attachments: list[dict] | None = Field(default=None, max_length=50)
    assigned_to_id: UUID | None = None
    related_equipment_id: UUID | None = None
    related_material_id: UUID | None = None

    @field_validator("subject", "question", "to_name", "location", mode="before")
    @classmethod
    def sanitize_text(cls, v: str | None) -> str | None:
        return sanitize_string(v)

    @field_validator("category", mode="before")
    @classmethod
    def validate_category(cls, v: str | None) -> str | None:
        if v is None:
            return v
        valid_categories = [
            "design", "structural", "mep", "architectural",
            "specifications", "schedule", "cost", "other"
        ]
        if v.lower() not in valid_categories:
            return "other"
        return v.lower()

    @field_validator("priority", mode="before")
    @classmethod
    def validate_priority(cls, v: str | None) -> str | None:
        if v is None:
            return v
        valid_priorities = ["low", "medium", "high", "urgent"]
        if v.lower() not in valid_priorities:
            return "medium"
        return v.lower()


class RFIStatusUpdate(BaseModel):
    status: str = Field(min_length=1, max_length=50)

    @field_validator("status", mode="before")
    @classmethod
    def validate_status(cls, v: str) -> str:
        valid_statuses = [
            "draft", "open", "waiting_response", "answered", "closed", "cancelled"
        ]
        if v and v.lower() not in valid_statuses:
            raise ValueError(f"Invalid status. Must be one of: {valid_statuses}")
        return v.lower()


class RFIResponseBase(BaseModel):
    response_text: str = Field(min_length=1, max_length=MAX_NOTES_LENGTH)
    attachments: list[dict] | None = Field(default=None, max_length=50)

    @field_validator("response_text", mode="before")
    @classmethod
    def sanitize_text(cls, v: str | None) -> str | None:
        return sanitize_string(v)


class RFIResponseCreate(RFIResponseBase):
    is_internal: bool = False


class RFIResponseSchema(BaseModel):
    id: UUID
    rfi_id: UUID
    email_message_id: str | None = None
    in_reply_to: str | None = None
    response_text: str
    attachments: list[dict] | None = None
    from_email: str
    from_name: str | None = None
    responder: UserResponse | None = None
    is_internal: bool = False
    is_cc_participant: bool = False
    source: str = "email"
    created_at: datetime
    received_at: datetime | None = None

    class Config:
        from_attributes = True


class RFIEmailLogSchema(BaseModel):
    id: UUID
    rfi_id: UUID | None = None
    event_type: str
    email_message_id: str | None = None
    from_email: str | None = None
    to_email: str | None = None
    subject: str | None = None
    error_message: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class RFIListResponse(BaseModel):
    id: UUID
    project_id: UUID
    rfi_number: str
    subject: str
    to_email: str
    to_name: str | None = None
    category: str
    priority: str
    status: str
    due_date: datetime | None = None
    created_at: datetime
    sent_at: datetime | None = None
    responded_at: datetime | None = None
    response_count: int = 0
    related_equipment_id: UUID | None = None
    related_material_id: UUID | None = None

    class Config:
        from_attributes = True


class RFIResponse(BaseModel):
    id: UUID
    project_id: UUID
    rfi_number: str
    email_thread_id: str | None = None
    email_message_id: str | None = None
    subject: str
    question: str
    category: str
    priority: str
    to_email: str
    to_name: str | None = None
    cc_emails: list[str] | None = None
    status: str
    due_date: datetime | None = None
    responded_at: datetime | None = None
    closed_at: datetime | None = None
    location: str | None = None
    drawing_reference: str | None = None
    specification_reference: str | None = None
    attachments: list[dict] | None = None
    created_at: datetime
    updated_at: datetime
    sent_at: datetime | None = None
    created_by: UserResponse | None = None
    assigned_to: UserResponse | None = None
    related_equipment_id: UUID | None = None
    related_material_id: UUID | None = None
    responses: list[RFIResponseSchema] = []

    class Config:
        from_attributes = True


class RFISummaryResponse(BaseModel):
    total_rfis: int
    draft_count: int
    open_count: int
    waiting_response_count: int
    answered_count: int
    closed_count: int
    overdue_count: int
    by_priority: dict[str, int]
    by_category: dict[str, int]


class RFISendRequest(BaseModel):
    send_now: bool = True


class PaginatedRFIResponse(BaseModel):
    items: list[RFIListResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class RFIDeadlineResponse(BaseModel):
    id: UUID
    project_id: UUID
    rfi_number: str
    subject: str
    to_email: str
    to_name: str | None = None
    category: str
    priority: str
    status: str
    due_date: datetime
    days_overdue: int | None = None
    days_until_due: int | None = None
    created_at: datetime
    sent_at: datetime | None = None

    class Config:
        from_attributes = True
